Model/Service/Index.php

```diff
@@ -0,0 +1,467 @@
+            ) {
+                $productIndex->setNostoProduct($nostoProduct);
+                $productIndex->setInSync(false);
+            }
+            $productIndex->setIsDirty(false);
+            $this->indexRepository->save($productIndex);
+            return $productIndex;
+        } catch (\Exception $e) {
+            $this->logger->exception($e);
+            return null;
+        }
+    }
+
+    /**
+     * Defines product index as in sync
+     *
+     * @param ProductIndexInterface $productIndex
+     * @throws \Exception
+     * @return void
+     */
+    public function markAsInSync(ProductIndexInterface $productIndex)
+    {
+        if (!$productIndex->getInSync()) {
+            $productIndex->setInSync(true);
+            $this->indexRepository->save($productIndex);
+        }
+    }
+
+    /**
+     * @param int $productId
+     * @param int $storeId
+     * @return void
+     * @throws \Exception
+     */
+    public function markAsInSyncProductByIdAndStore($productId, $storeId)
+    {
+        try {
+            $productIndex = $this->indexRepository->getByProductIdAndStoreId($productId, $storeId);
+            if ($productIndex instanceof ProductIndexInterface) {
+                $this->markAsInSync($productIndex);
+            }
+        } catch (\Exception $e) {
+            $this->logger->exception($e);
+        }
+    }
+
+    /**
+     * @param ProductCollection $collection
+     * @param array $ids
+     * @param Store $store
+     */
+    public function markProductsAsDeletedByDiff(ProductCollection $collection, array $ids, Store $store)
+    {
+        $uniqueIds = array_unique($ids);
+        $collection->setPageSize(self::PRODUCT_DELETION_BATCH_SIZE);
+        $lastPage = $collection->getLastPageNumber();
+        $curPage = 1;
+        do {
+            $collection->clear();
+            $collection->setCurPage($curPage);
+            /** @var Product $magentoProduct */
+            foreach ($collection as $magentoProduct) {
+                $key = array_search($magentoProduct->getId(), $uniqueIds);
+                if (is_numeric($key)) {
+                    unset($uniqueIds[$key]);
+                }
+            }
+            ++$curPage;
+        } while ($curPage <= $lastPage);
+
+        // Flag the rest of the ids as deleted
+        $deleted = $this->nostoIndexCollectionFactory->create()->markAsDeleted($uniqueIds, $store);
+        $this->logger->info(
+            sprintf(
+                'Marked %d records as deleted for store %s',
+                $deleted,
+                $store->getName()
+            )
+        );
+    }
+
+    /**
+     * Discontinues products in Nosto and removes indexed products from Nosto product index
+     *
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @return int number of deleted products
+     * @throws NostoException
+     */
+    public function handleProductDeletion(NostoIndexCollection $collection, Store $store)
```

Comment:
This entire method needs to be refactored. I think that this is a violation of SRP and needs to be modularised. We should do this as the last change.